syncer/syncer.go

```diff
@@ -0,0 +1,419 @@
+		var err error
+		if layer, err = s.syncLayer(ctx, layerID); err != nil {
+			logger.With().Error("failed to sync to layer", log.Err(err))
+			return false
+		}
+
+		if len(layer.Blocks()) == 0 {
+			logger.Info("setting layer %v to zero-block", layerID)
+			if err := s.mesh.SetZeroBlockLayer(layerID); err != nil {
+				logger.With().Error("failed to set zero-block for layer", layerID, log.Err(err))
+				return false
+			}
+		}
+
+		if s.shouldValidateLayer(layerID) {
+			vQueue <- layer
+		}
+		logger.Info("finished data sync for layer %v", layerID)
+	}
+	logger.With().Info("data is synced. waiting for validation",
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+	success = true
+	return true
+}
+
+func (s *Syncer) setStateBeforeSync(ctx context.Context) {
+	current := s.ticker.GetCurrentLayer()
+	if current == 0 {
+		s.setSyncState(ctx, synced)
+	}
+	latest := s.mesh.LatestLayer()
+	if current > latest && current-latest >= outOfSyncThreshold {
+		s.logger.WithContext(ctx).With().Info("node is too far behind",
+			log.FieldNamed("current", current),
+			log.FieldNamed("latest", latest),
+			log.FieldNamed("behindThreshold", outOfSyncThreshold))
+		s.setSyncState(ctx, notSynced)
+	}
+}
+
+func (s *Syncer) setStateAfterSync(ctx context.Context, success bool) {
+	if !success {
+		s.setSyncState(ctx, notSynced)
+		return
+	}
+	currSyncState := s.getSyncState()
+	current := s.ticker.GetCurrentLayer()
+	// if we have gossip-synced to the target synced layer, we are ready to participate in the consensus
+	if currSyncState == gossipSync && s.getTargetSyncedLayer() <= current {
+		s.setSyncState(ctx, synced)
+	} else if currSyncState == notSynced {
+		// wait till s.ticker.GetCurrentLayer() + numGossipSyncLayers to participate in consensus
+		s.setSyncState(ctx, gossipSync)
+		s.setTargetSyncedLayer(ctx, current+numGossipSyncLayers)
+	}
+}
+
+func (s *Syncer) syncLayer(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	if s.isClosed() {
+		return nil, errors.New("shutdown")
+	}
+
+	layer, err := s.getLayerFromPeers(ctx, layerID)
+	if err != nil {
+		s.logger.WithContext(ctx).With().Error("failed to get layer from neighbors", layerID, log.Err(err))
+		return nil, err
+	}
+
+	if err := s.getATXs(ctx, layerID); err != nil {
+		return nil, err
+	}
+	return layer, nil
+}
+
+func (s *Syncer) getLayerFromPeers(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	ch := s.fetcher.PollLayer(ctx, layerID)
+	res := <-ch
+	if res.Err != nil {
+		if res.Err == layerfetcher.ErrZeroLayer {
+			return types.NewLayer(layerID), nil
+		}
+		return nil, res.Err
+	}
+	return s.mesh.GetLayer(layerID)
+}
+
+func (s *Syncer) getATXs(ctx context.Context, layerID types.LayerID) error {
+	if layerID.GetEpoch() == 0 {
+		s.logger.WithContext(ctx).Info("skip getting ATX in epoch 0")
+		return nil
+	}
+	epoch := layerID.GetEpoch()
```

Comment:
now we get ATXs for each layer of an epoch, we can however optimise this, if we are syncing an epoch that is not the current epoch we can sync only once
If we are in the current epoch we will need to query for each layer or try to sync atxs when we sync the last layer we need to sync in current epoch. - if implemented this way - be careful since syncing atxs might take a while

Reply:
changed to what you suggested.
- get ATXs in the last layer of an old epoch
- get ATXs for every layer of current epoch to amortize the latency